### ISSUE TITLE
GRANT problems

### ISSUE THREAD

**Issue by walter-weinmann** [K2InformaticsGmbH/sqlparse#22]
The following GRANT statements are not accepted by the parser:

```
GRANT ON DIRECTORY name_directory TO name_user_1
GRANT ON JAVA SOURCE name_table TO name_user_1
GRANT ON JAVA RESOURCE name_table TO name_user_1
```

What is the use of a GRANT statement without system_privilege_list?

**Comment by c-bik** [comment]
As per [RRDs](http://docs.oracle.com/cd/B19306_01/server.102/b14200/statements_9013.htm) the unaccepted grant statements you listed above doesn't look valid to me. What am I missing?

![image](https://cloud.githubusercontent.com/assets/913008/13858340/829e812e-ec7e-11e5-8045-fa3891683b28.png)

grant_system_privileges::=
![image](https://cloud.githubusercontent.com/assets/913008/13858343/88372230-ec7e-11e5-950c-1a2769fd4309.png)

grant_object_privileges::=
![image](https://cloud.githubusercontent.com/assets/913008/13858364/9fc95e86-ec7e-11e5-94a2-927cd6b9d22d.png)

on_object_clause ::=
![image](https://cloud.githubusercontent.com/assets/913008/13858375/aeb82ec2-ec7e-11e5-8f23-abda1032c353.png)

grantee_clause ::=
![image](https://cloud.githubusercontent.com/assets/913008/13858385/b79744ce-ec7e-11e5-8a9f-c9b9c80adec5.png)

**Comment by walter-weinmann** [comment]
The test cases:

GRANT ON DIRECTORY name_directory TO name_user_1
GRANT ON JAVA SOURCE name_table TO name_user_1
GRANT ON JAVA RESOURCE name_table TO name_user_1

are based on the following grammar rules:

grant_def ->
 GRANT system_privilege_list opt_on_obj_clause TO grantee_commalist
opt_with_grant_option       : {'grant', '$2', '$3', {'to', '$5'},
'$6'}.

opt_on_obj_clause -> '$empty'
                         : {'on', <<"">>}.
opt_on_obj_clause -> ON table
                         : {'on', '$2'}.
opt_on_obj_clause -> ON DIRECTORY NAME
                         : {'on directory', unwrap_bin('$3')}.
opt_on_obj_clause -> ON JAVA SOURCE table
                         : {'on java source', unwrap_bin('$4')}.
opt_on_obj_clause -> ON JAVA RESOURCE table
                         : {'on java resource', unwrap_bin('$4')}.

system_privilege_list -> '$empty'
                         : [].
system_privilege_list -> system_privilege
                         : ['$1'].
system_privilege_list -> system_privilege ',' system_privilege_list
                         : ['$1'|'$3'].
system_privilege_list -> ALL
                         : ['all'].
system_privilege_list -> ALL PRIVILEGES
                         : ['all privileges'].



On 17 March 2016 at 20:31, Bikram Chatterjee <notifications@github.com>
wrote:

> As per RRDs
> <http://docs.oracle.com/cd/B19306_01/server.102/b14200/statements_9013.htm>
> the unaccepted grant statements you listed above doesn't look valid to me.
> What am I missing?
>
> [image: image]
> <https://cloud.githubusercontent.com/assets/913008/13858340/829e812e-ec7e-11e5-8045-fa3891683b28.png>
>
> grant_system_privileges::=
> [image: image]
> <https://cloud.githubusercontent.com/assets/913008/13858343/88372230-ec7e-11e5-950c-1a2769fd4309.png>
>
> grant_object_privileges::=
> [image: image]
> <https://cloud.githubusercontent.com/assets/913008/13858364/9fc95e86-ec7e-11e5-94a2-927cd6b9d22d.png>
>
> on_object_clause ::=
> [image: image]
> <https://cloud.githubusercontent.com/assets/913008/13858375/aeb82ec2-ec7e-11e5-8f23-abda1032c353.png>
>
> grantee_clause ::=
> [image: image]
> <https://cloud.githubusercontent.com/assets/913008/13858385/b79744ce-ec7e-11e5-8a9f-c9b9c80adec5.png>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/K2InformaticsGmbH/sqlparse/issues/22#issuecomment-198047409>
>



-- 

___________________________________________________________________________
*Walter Weinmann*
___________________________________________________________________________
Obertorplatz 4
CH-4130 Rheinfelden
Tel +41 (0)61 841 06 10
___________________________________________________________________________
Schulstrasse 1
CH-6037 Root
Tel +41 (0)41 530 39 70
___________________________________________________________________________

**Comment by c-bik** [comment]
> What is the use of a GRANT statement without system_privilege_list?
`system_privilege_list -> '$empty'`

Starting to see the problem... that rule is incorrect... will cross check again with Oracle ref

**Comment by walter-weinmann** [comment]
Suggested solution: remove the following grammar rules:

```
grant_def -> GRANT                                     TO grantee_commalist                     : {grant, [],   {on, <<"">>}, {to, '$3'}, ''}.
grant_def -> GRANT                                     TO grantee_commalist with_grant_option   : {grant, [],   {on, <<"">>}, {to, '$3'}, '$4'}.
grant_def -> GRANT                       on_obj_clause TO grantee_commalist                     : {grant, [],   '$2',         {to, '$4'}, ''}.
grant_def -> GRANT                       on_obj_clause TO grantee_commalist with_grant_option   : {grant, [],   '$2',         {to, '$4'}, '$5'}.

```

**Comment by walter-weinmann** [comment]
Solved with #82.
